### ISSUE TITLE
Question about Clustering

### ISSUE THREAD

**Issue by PhilipAmadasun** [ethz-asl/polygon_coverage_planning#73]
How is the clustering done? How does the algorithm know how many clusters to make . I would assume it has to do with the size and shape of of the "no fly" zones as well as that of the boundaries of the map. What is the technique though?

**Comment by rikba** [comment]
You are correct, the shape of the decomposition depends a lot on the shape of the polygon and holes. That's why this approach is not the best if you have many, complex obstacles, because you will create many cells.

You can [pick](https://github.com/ethz-asl/polygon_coverage_planning/blob/master/polygon_coverage_ros/cfg/coverage_planner.yaml#L1) between Boustrophedon decomposition or trapezoidal decomposition. These are so called [sweep line algorithms](https://en.wikipedia.org/wiki/Sweep_line_algorithm). The [trapezoidal decomposition is part of CGAL](https://github.com/ethz-asl/polygon_coverage_planning/blob/master/polygon_coverage_geometry/src/tcd.cc#L36-L37), while the [Boustrophedon decomposition we implemented ourselves](https://github.com/ethz-asl/polygon_coverage_planning/blob/master/polygon_coverage_geometry/src/bcd.cc). Because they are sweep line algorithm the resulting decomposition depends on the direction of the sweep. We iterate through all sweep directions perpendicular to any of the edge directions and select the one that has the [minimum altitude sum](https://github.com/ethz-asl/polygon_coverage_planning/blob/master/polygon_coverage_geometry/src/decomposition.cc#L118). 

If you want to read more on this you can have a look at the following publications:

Choset, H., Pignon, P.: Coverage path planning: The boustrophedon cellular decomposition. In: Field and Service Robotics, pp. 203–209. Springer (1998)

Huang, W.H.: Optimal line-sweep-based decompositions for coverage algorithms. In:
IEEE International Conference on Robotics and Automation, vol. 1, pp. 27–32. IEEE
(2001)

as well as this presentation: https://www.cs.cmu.edu/~motionplanning/lecture/Chap6-CellDecomp_howie.pdf
